**DAT405 Introduction to Data Science and AI, 2010-2021, Study Period 2** <br/>
**Assignment 5: Reinforcement learning and Classification** <br/>
**Due Date: Feb 24, 23:59** <br/>

**Group 6:**

Name | Contribution 
--- | ---
1. Himanshu Chuphal (guschuhi@student.gu.se) | 12 H
2. Claudio Aguilar Aguilar(claagu@student.chalmers.se) | 12 H

---
---


**What to submit**
*   **The entire assignment should be submitted through the notebook. No separate file will be accepted.** You can submit either the notebook itself, or a public link to a Google Colab notebook<br/>

*In the notebook:*
*	State your names and how many hours each person spent on the assignment.
*	The solutions and answers to the theoretical and practical problems, including LaTeX math-mode equations, plots and tables etc.
*	All plots/results should be visible such that the notebook does not have to be run. But the code in the notebook should reproduce the plots/results if we choose to do so.<br/>

*Before submitting:*
*   Make sure that your code can run on another computer. That all plots can show on another computer including all your writing. It is good to check if your code can run here: https://colab.research.google.com.

**Self-check**<br/>
Is all the required information included? Have you answered all questions to the best of your ability? Anything else you can easily check? (details, terminology, arguments, clearly stated answers etc.?) Does your notebook run and can reproduce the results, plots and tables?

**Grading**<br/>
Grading will be based on a qualitative assessment of each assignment. It is important to:
*	Present clear arguments
*	Present the results in a pedagogical way
*	Show understanding of the topics (e.g, write a pseudocode) 
*	Give correct solutions
*	Make sure that the code is well commented 

**Again, as mentioned in general guidelines, all code should be written here. And this same ipython notebook file (Assignment5_Reinforcement_Learning.ipynb) should be submitted with answers and code written in it. No separate file will be accepted.** 


# Primer

## Decision Making
The problem of **decision making under uncertainty** (commonly known as **reinforcement learning**) can be broken down into
two parts. First, how do we learn about the world? This involves both the
problem of modeling our initial uncertainty about the world, and that of drawing conclusions from evidence and our initial belief. Secondly, given what we
currently know about the world, how should we decide what to do, taking into
account future events and observations that may change our conclusions?
Typically, this will involve creating long-term plans covering possible future
eventualities. That is, when planning under uncertainty, we also need to take
into account what possible future knowledge could be generated when implementing our plans. Intuitively, executing plans which involve trying out new
things should give more information, but it is hard to tell whether this information will be beneficial. The choice between doing something which is already
known to produce good results and experiment with something new is known
as the **exploration-exploitation dilemma**.

## The exploration-exploitation trade-off

Consider the problem of selecting a restaurant to go to during a vacation. Lets say the
best restaurant you have found so far was **Les Epinards**. The food there is
usually to your taste and satisfactory. However, a well-known recommendations
website suggests that **King’s Arm** is really good! It is tempting to try it out. But
there is a risk involved. It may turn out to be much worse than **Les Epinards**,
in which case you will regret going there. On the other hand, it could also be
much better. What should you do?
It all depends on how much information you have about either restaurant,
and how many more days you’ll stay in town. If this is your last day, then it’s
probably a better idea to go to **Les Epinards**, unless you are expecting **King’s
Arm** to be significantly better. However, if you are going to stay there longer,
trying out **King’s Arm** is a good bet. If you are lucky, you will be getting much
better food for the remaining time, while otherwise you will have missed only
one good meal out of many, making the potential risk quite small.

## Overview
* To make things concrete, we will first focus on decision making under **no** uncertainity, i.e, given we have a world model, we can calculate the exact and optimal actions to take in it. We will first introduce **Markov Decision Process (MDP)** as the world model. Then we give one algorithm (out of many) to solve it.


* Next, we will work through one type of reinforcement learning algorithm called Q-learning. Q-learning is an algorithm for making decisions under uncertainity, where uncertainity is over the possible world model (here MDP). It will find the optimal policy for the **unknown** MDP, assuming we do infinite exploration.

## Markov Decision Process

Markov Decision Process (MDP) provides a mathematical framework for modeling sequential decision making under uncertainty. A MDP consists of five parts: the specific decision times, the state space of the environment/system, the available actions for the decision maker, the rewards, and the transition probabilities between the states.

* Decision epochs: $t={1,2,...,T}$, where $T\leq \infty$
* State space: $S=\{s_1,s_2,...,s_N\}$ of the underlying environment
* Action space $A=\{a_1,a_2,...,a_K\}$ available to the decision maker at each decision epoch
* Reward functions $R_t = r(a_t,s_t,s_{t+1})$ for the current state and action, and the resulting next state
* Transition probabilities $p(s'|s,a)$ that taking action $a$ in state $s$ will lead to state $s'$

At a given decision epoch $t$ and system state $s_t$, the decions maker, or *agent*, chooses an action $a_t$, the system jumps to a new state $s_{t+1}$ according to the transition probability $p(s_{t+1}|s_t,a_t)$, and the agent receives a reward $r_t(s_t,a_t,s_{t+1})$. This process is then repeated for a finite or infinite number of times.

A *decision policy* is a function $\pi: s \rightarrow a$, that gives instructions on what action to choose in each state. A policy can either be *deterministic*, meaning that the action is given for each state, or *randomized* meaning that there is a probability distribution over the set of possible actions. Given a specific policy $\pi$ we can then compute the the *expected total reward* when starting in a given state $s_1 \in S$, which is also known as the *value* for that state, 

$$V^\pi (s_1) = E\left[ \sum_{t=1}^{T} r(s_t,a_t,s_{t+1}) {\Large |} s_1\right] = \sum_{t=1}^{T} r(s_t,a_t,s_{t+1}) p(s_{t+1} | a_t,s_t)$$ 

where $a_t = \pi(s_t)$. To ensure convergence and to control how much credit to give to future rewards, it is common to introduce a *discount factor* $\gamma \in [0,1]$. For instance, if you think all future rewards should count equally, you would use $\gamma = 1$, while if you only care less about future rewards you would use $\gamma < 1$. The expected total *discounted* reward becomes

$$V^\pi( s_1) = \sum_{t=1}^T \gamma^{t-1} r(s_t,a_t, s_{t+1}) p(s_{t+1} | s_t, a_t) $$

Now, to find the *optimal* policy we want to find the policy $\pi^*$ that gives the highest total reward $V^{\pi^*}(s)$ for all $s\in S$. That is

$$V^{\pi^*}(s) \geq V^\pi(s), s\in S$$

The problem of finding the optimal policy is a _dynamic programming problem_. It turns out that a solution to the optimal policy problem in this context is the *Bellman equation*. The Bellman equation is given by

$$V(s) = \max_{a\in A} \left\{\sum_{s'\in S} p(s'|s,a)( r(s,a,s') +\gamma V(s')) \right\}$$

Thus, it can be shown that if $\pi$ is a policy such that $V^\pi$ fulfills the Bellman equation, then $\pi$ is an optimal policy.

A real world example would be an inventory control system. Your states would be the amount of items you have in stock. Your actions would be the amount to order. The discrete time would be the days of the month. The reward would be the profit.  

A major drawback of MDPs is called the "Curse of Dimensionality". MDPs unfortunately do not scale very well with increasing sets of states or actions.   


## Question 1

In this first question we work with the deterministic MDP, no code is necessary in this part.

Setup:

* The agent starts in state **S**
* The actions possible are **N** (north), **S** (south), **E** (east), and **W** west. 
* Note, that you cannot move outside the grid, thus all actions are not available in every box.
* When reaching **F**, the game ends (absorbing state).
* The numbers in the boxes represent the rewards you receive when moving into that box. 
* Assume no discount in this model: $\gamma = 1$

The reward of a state $r(s=(x, y))$ is given by the values on the grid:
    
| | | |
|----------|----------|---------|
|-1 |1|**F**|
|0|-1|1|  
|-1 |0|-1|  
|**S**|-1|1|

Let $(x,y)$ denote the position in the grid, such that $S=(0,0)$ and $F=(2,3)$.

**1a)** What is the optimal path of the MDP above? Is it unique? Submit the path as a single string of directions. E.g. NESW will make a circle.

**[Answer]**
The optimal path is **EENNN** since it is unique in the way that it is the shortest path with the sum of 0.

But there is also another the path **EENNWNE** that results in a sum of 0 but takes 2 additional steps to do so.


**1b)** What is the optimal policy (i.e. the optimal action in each state)?

**[Answer]**
The optimal policy to complete the optimal path is:

1.   When in State(0,0) do E
2.   When in State(1,0) do E
3.   When in State(2,0) do N
4.   When in State(2,1) do N
5.   When in State(2,2) do N

All other alternative paths until we reach the goal::

**(0,0)**: E or N = -1, **(0,1)**: E = 1, **(0,2)** : W or N = -1

**(1,0)**: E or N = 0 (return to start S is also possible), **(1,1)**: E or W or N or S = -1, **(1,2)**: N or S = 1 

**(2,0)**: E or N or S =-1 , **(2,1)**: E or N = 1, **(2,2)**: N = 0

**(3,0)**:  E = 1, **(3,1)**: E = 0, **(3,2)**: No possible action!

**1c)** What is expected total reward for the policy in 1b)?

**[Answer]**
The result that the optimal path **EENNN** sums up to, which is **0**.
Here, V(s) is the total sum value we expect to obtain when you follow a current policy starting from initial state (s), which is calculated using following formula: 

- new value from an action = (currentStateReward + 1 * nextStateReward) 
*with gamma =1*

 - V(s) =  expected total reward (added up values for each action)



(0,0):N, (1,0):E, (2,0):N, (2,1):N, (2,2):N , (3,2):: **SUM** =-1+0+0+0+1 = 0

optimal path from all other alternative paths  from 1(b) ::

(0,0):N, (1,0):E, (1,1):E, (1,2):N, (2,2):N, (3,2) :: **sum** = -1-1-1+0+1= -2

(0,0):N, (1,0):E, (1,1):N, (2,1):E, (2,2):N, (3,2) :: **sum** = -1-1-1+0+1= -2

(0,0):N, (1,0):E, (1,1):N, (2,1):N, (3,1):E, (3,2) :: **sum** = -1-1-1+0+1= -2

(0,0):N, (1,0):N, (2,0):E, (2,1):E, (2,2):N, (3,2) :: **sum** = -1-1-1+0+1= -2

(0,0):N, (1,0):N, (2,0):E, (2,1):N, (3,1):E, (3,2) :: **sum** = -1-1-1+0+1= -2

(0,0):N, (1,0):N, (2,0):N, (3,0):E, (3,1):E, (3,2) :: **sum** = -1-1-1+0+1= -2

The other possible paths have worse value.


## Value Iteration

For larger problems we need to utilize algorithms to determine the optimal policy $\pi^*$. *Value iteration* is one such algorithm that iteratively computes the value for each state. Recall that for a policy to be optimal, it must satisfy the Bellman equation above, meaning that plugging in a given candidate $V^*$ in the right-hand side (RHS) of the Bellman equation should result in the same $V^*$ on the left-hand side (LHS). This property will form the basis of our algorithm. Essentially, it can be shown that repeated application of the RHS to any intial value function $V^0(s)$ will eventually lead to the value $V$ which statifies the Bellman equation. Hence repeated application of the Bellman equation will also lead to the optimal value function. We can then extract the optimal policy by simply noting what actions that satisfy the equation. The process of repeated application of the Bellman equation what we here call the _value iteration_ algorithm.

The value iteration algorithm practically procedes as follows:

```
epsilon is a small value, threshold
for x from i to infinity 
do
    for each state s
    do
        V_k[s] = max_a Σ_s' p(s′|s,a)*(r(a,s,s′) + γ*V_k−1[s′])
    end
    if  |V_k[s]-V_k-1[s]| < epsilon for all s
        for each state s,
        do
            π(s)=argmax_a ∑_s′ p(s′|s,a)*(r(a,s,s′) + γ*V_k−1[s′])
            return π, V_k 
        end
end

```






**Example:** We will illustrate the value iteration algorithm by going through two iterations. Below is a 3x3 grid with the rewards given in each state. Assume now that given a certain state $s$ and action $a$, there is a probability of 0.8 that that action will be performed and a probability of 0.2 that no action is taken. For instance, if we take action **E** in state $(x,y)$ we will go to $(x+1,y)$ 80 percent of the time (given that that action is available in that state, that is, we stay on the grid), and remain still 20 percent of the time. We will use have a discount factor $\gamma = 0.9$. Let the initial value be $V^0(s)=0$ for all states $s\in S$. 

| | | |  
|----------|----------|---------|  
|0|0|0|
|0|10|0|  
|0|0|0|  


**Iteration 1**: The first iteration is trivial, $V^1(s)$ becomes the $\max_a \sum_{s'} p(s'|s,a) r(s,a,s')$ since $V^0$ was zero for all $s'$. The updated values for each state become

| | | |  
|----------|----------|---------|  
|0|8|0|
|8|2|8|  
|0|8|0|  
  
**Iteration 2**:  
  
Staring with cell (0,0) (lower left corner): We find the expected value of each move:  
Action **S**: 0  
Action **E**: 0.8( 0 + 0.9 \* 8) + 0.2(0 + 0.9 \* 0) = 5.76  
Action **N**: 0.8( 0 + 0.9 \* 8) + 0.2(0 + 0.9 \* 0) = 5.76  
Action **W**: 0

Hence any action between **E** and **N** would be best at this stage.

Similarly for cell (1,0):

Action **N**: 0.8( 10 + 0.9 \* 2) + 0.2(0 + 0.9 \* 8) = 10.88 (Action **N** is the maximizing action)  

Similar calculations for remaining cells give us:

| | | |  
|----------|----------|---------|  
|5.76|10.88|5.76|
|10.88|8.12|10.88|  
|5.76|10.88|5.76|  


In [1]:
#imported python modules
import numpy as np
import gym

## Question 2

**2a)** Implement the value iteration algorithm just described here in python, and show the converging optimal value function and the optimal policy for the above 3x3 grid. Hint: use the pseudo-code above as a starting point, but be sure to explain what every line does.

**converging optimal value function**

$$V^\pi( s_1) = \sum_{t=1}^T \gamma^{t-1} r(s_t,a_t, s_{t+1}) p(s_{t+1} | s_t, a_t) $$



**optimal policy (3x3 grid, fulfills bellman equation)**

$$V(s) = \max_{a\in A} \left\{\sum_{s'\in S} p(s'|s,a)( r(s,a,s') +\gamma V(s')) \right\}$$





In [2]:
import numpy as np
#initial and reward matrix
initial_reward =  np.arange(1.0, 10.0).reshape(3,3)
reward_matrix =  np.arange(1.0, 10.0).reshape(3,3)
grid = np.arange(1.0,10.0).reshape(3,3)
#probability of 0.8 that that action will be performed and a probability of 0.2 that no action is taken
probAction, probNotAction, Gamma = 0.8, 0.2, 0.9
#zero the grid
def get_inital_grid():
  for i in range (0,3):
    for j in range(0,3):
      grid[i][j] = 0.0
  return grid
  
print("Initial Grid :")
print(get_inital_grid())

#3x3 grid.
for i in range(0,3):
    for j in range (0,3):
        initial_reward[i][j] = 0.0
initial_reward[1][1]=10.0
print("Initial Reward :")
print(initial_reward)

#size of grid (width x height)
grid_width  = grid.shape[0]-1
grid_height = grid.shape[1]-1

def set_action(grid, x, y, action):
  #Define the different actions
  #EAST
  if (action == 'E'):
    if (x+1 <= grid_width and x+1 >= 0): #if within the grid then step to the right
      return grid[x+1][y]
    else: 
      return 0 #action not possible
  #WEST
  if (action == 'W'):
    if (x-1 <= grid_width and x-1 >= 0): #if within the grid and >= 0 then step to the left
      return grid[x-1][y]
    else:
      return 0 #action not possible
  #SOUTH
  if (action == 'S'):
    if (y-1 <= grid_height and y-1 >= 0): #if within the grid and >= 0 then step down
      return grid[x][y-1]
    else:
      return 0 #action not possible
  #NORTH
  if (action == 'N'):
    if (y+1 <= grid_height and y+1 >= 0): #if within the grid and >= 0 then step up
      return grid[x][y+1]
    else:
      return 0 #action not possible
  #OLDVALUE
  if (action == 'O'):
    return grid[x][y]

#max value as the bellman equation implies as stated in question 2
def max_value_action(x, y, current, next):
    N = probAction * (set_action(current,x,y,'N')+ Gamma * set_action(next ,x,y,'N'))+ probNotAction * (current[x][y]+ Gamma * next[x][y])
    S = probAction * (set_action(current,x,y,'S')+ Gamma * set_action(next ,x,y,'S'))+ probNotAction * (current[x][y]+ Gamma * next[x][y])
    E = probAction * (set_action(current,x,y,'E')+ Gamma * set_action(next ,x,y,'E'))+ probNotAction * (current[x][y]+ Gamma * next[x][y])
    W = probAction * (set_action(current,x,y,'W')+ Gamma * set_action(next ,x,y,'W'))+ probNotAction * (current[x][y]+ Gamma * next[x][y])
    return (max(N,S,E,W))

#expected reward for each state in one iteration
for i in range(0,3):
  for j in range (0,3):
    reward_matrix[i][j] = max_value_action(i, j, initial_reward, get_inital_grid())
print("After 1st Iteration::")
print(reward_matrix)

#start the value iteration for total_iteration times
def ValueIteration(total_iteration):
  final_grid = get_inital_grid() 
  reward = reward_matrix #expected reward from last iteration
  for _ in range(total_iteration-1):
    for i in range(0,3):
      for j in range(0,3):
        final_grid[i][j] = max_value_action(i, j, initial_reward, reward)
    reward = final_grid 
  return final_grid 

final_grid = ValueIteration(1000)
print("Value after 1000 Interations ::")
print(final_grid)

Initial Grid :
[[0. 0. 0.]
 [0. 0. 0.]
 [0. 0. 0.]]
Initial Reward :
[[ 0.  0.  0.]
 [ 0. 10.  0.]
 [ 0.  0.  0.]]
After 1st Iteration::
[[0. 8. 0.]
 [8. 2. 8.]
 [0. 8. 0.]]
Value after 1000 Interations ::
[[45.61292366 51.94805195 45.61292366]
 [51.94805195 48.05194805 51.94805195]
 [45.61292366 51.94805195 45.61292366]]


**2b)** Explain why the result of 2a) does not depend on the initial value $V_0$.

**[Answer]** It is probably because every single state is handeled separately. and applicable actions for each states are independent of each other. Moreover. the initial value $V_0$ consist of small initial reward which will spread out after the 1st iteration. Incase, we have comparetively bigger number in the inital values, then also it would not matter if we increase the total number of iterations to converge the values. The oldest rewards values i.e. the initial value $V_0$ will be taken into account the least after large number of iterations. Beacause of which impact of inital values will be very less.

## Reinforcement Learning (RL)
Until now, we understood that knowing the MDP, specifically $p(s'|a,s)$ and $r(a,s,s')$ allows us to efficiently find the optimal policy using the value iteration algorithm. Reinforcement learning (RL) or decision making under uncertainity, however, arises from the question of making optimal decisions without knowing the true world model (the MDP in this case).

So far we have defined the value function for a policy through $V^\pi$. Let's now define the *action-value function*

$$Q^\pi(s,a) = \sum_{s'} p(s'|a,s) [r(a,s,s') + \gamma V^\pi(s')]$$

The value function and the action-value function are directly related through

$$V^\pi (s) = \max_a Q^\pi (s,a)$$

i.e, the value of taking action $a$ in state $s$ and then following the policy $\pi$ onwards. Similarly to the value function, the optimal $Q$-value equation is:

$$Q^*(s,a) = \sum_{s'} p(s'|a,s) [r(a,s
]\,s') + \gamma V^*(s')]$$

and the relationship between $Q^*(s,a)$ and $V^*(s)$ is simply

$$V^*(s) = \max_{a\in A} Q^*(s,a).$$

## Q-learning

Q-learning is a RL-method where the agent learns about its unknown environment (i.e. the MDP is unknown) through exploration. In each time step *t* the agent chooses an action *a* based on the current state *s*, observes the reward *r* and the next state *s'*, and repeats the process in the new state. Q-learning is then a method that allows the agent to act optimally. Here we will focus on the simplest form of Q-learning algorithms, which can be applied when all states are known to the agent, and the state and action spaces are reasonably small. This simple algorithm uses a table of Q-values for each $(s,a)$ pair, which is then updated in each time step using the update rule in step $k+1$

$$Q_{k+1}(s,a) = Q_k(s,a) + \alpha \left( r(s,a) + \gamma \max \{Q_k(s',a')\} - Q_k(s,a) \right) $$ 

where $\gamma$ is the discount factor as before, and $\alpha$ is a pre-set learning rate. It can be shown that this algorithm converges to the optimal policy of the underlying MDP for certain values of $\alpha$ as long as there is sufficient exploration. While a constant $\alpha$ generally does not guarantee us to reach true convergence, we keep it constant at $\alpha=0.1$ for this assignment.

## OpenAI Gym

We shall use already available simulators for different environments (worlds) using the popular OpenAI Gym library. It just implements [different types of simulators](https://gym.openai.com/) including ATARI games. Although here we will only focus on simple ones, such as the [Chain enviroment](https://gym.openai.com/envs/NChain-v0/) illustrated below.
![alt text](https://chalmersuniversity.box.com/shared/static/6tthbzhpofq9gzlowhr3w8if0xvyxb2b.jpg)
The figure corresponds to an MDP with 5 states $S = \{1,2,3,4,5\}$ and two possible actions $A=\{a,b\}$ in each state. The arrows indicate the resulting transitions for each state-action pair, and the numbers correspond to the rewards for each transition.

## Question 3
You are to first familiarize with the framework using its [documentation](http://gym.openai.com/docs/), and then implement the Q-learning algorithm for the Chain enviroment (called 'NChain-v0') using default parameters. Finally print the $Q^*$ table at convergence. Convergence is **not** a constant value, rather a stable plateau with some noise. Take $\gamma=0.95$. You can refer to the Q-learning (frozen lake) Jupyter notebook shown in class, uploaded on Canvas. Hint: start with a small learning rate.




In [3]:
!apt-get update > /dev/null 2>&1
!pip install gym pyvirtualdisplay > /dev/null 2>&1
!apt-get install -y xvfb python-opengl ffmpeg > /dev/null

Q-learning is one of the most famous model-free reinforcement learning algorithms. It aims to find the optimal policy by maximizing the expected value of the total rewards from the current step to all the following steps. In this sense, Q in Q-learning stands for "quality", meaning that the algorithm aims to find the policy with the best quality for the agent. It is described by the Bellman equation:

In [4]:
#imported python modules
import gym
import random
import numpy as np 
from IPython.display import Math, HTML
#display(HTML("<script src='https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.3/"
#               "latest.js?config=default'></script>"))

Math(r"NewQ(s,a) = Q(s,a) + \alpha [R(s,a) + \gamma max Q(s',a')-Q(s,a)]")

<IPython.core.display.Math object>

In [10]:
# taken from q_learning_frozen_lake.ipynb
env = gym.make('NChain-v0')
num_episodes =  1000#15000 #20000 #60000
gamma = 0.95 #0.99
learning_rate = 0.1 #0.95 #0.85
epsilon = 0.5#1 #0.15 #0.1

# initialize the Q table
Q = np.zeros([5, 2])
Q

array([[0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.],
       [0., 0.]])

In [11]:
## taken from q_learning_frozen_lake.ipynb
for _ in range(num_episodes):
	state = env.reset()
	done = False
	while done == False:
        # First we select an action:
		if random.uniform(0, 1) < epsilon: # Flip a skewed coin
			action = env.action_space.sample() # Explore action space
		else:
			action = np.argmax(Q[state,:]) # Exploit learned values
        # Then we perform the action and receive the feedback from the environment
		new_state, reward, done, info = env.step(action)
        # Finally we learn from the experience by updating the Q-value of the selected action
		update = reward + (gamma*np.max(Q[new_state,:])) - Q[state, action]
		Q[state,action] += learning_rate*update 
		state = new_state
#Printing the Q table
Q

array([[62.73119814, 61.61184087],
       [66.77077242, 61.8414113 ],
       [71.61108039, 62.89457625],
       [76.54452395, 62.27858727],
       [81.55990703, 67.58229107]])

## Question 4

**4a)** Define the MDP corresponding to the Chain environment above and verify that the optimal $Q^*$ value obtained using simple Q-learning is the same as the optimal value function $V^*$ for the corresponding MDP's optimal action. Hint: compare values obtained using value iteration and Q-learning.


In [7]:
#Using Figure 1 states and rewards 
states = [1,2,3,4,5] #total 5 states
reward_b = 2. # for backward
reward_f = 0. # for forward
reward_f5a = 10. #for state 5
gamma = 0.95 # as given in the previous question 3
#expected reward for each state in one iteration
initial_val = [0.0, 0.0, 0.0, 0.0, 0.0]

#updating max value action from question 2
def max_value_action_2(state, next): # based on Figure 1
    # for states 1 to 4
    if(state >= 0 and state <4): 
      forward_reward = probAction * (reward_f + gamma * next[state+1])+ probNotAction * (reward_b + gamma * next[0])
      backward_reward = probAction * (reward_b + gamma * next[0])+ probNotAction * (gamma * next[state+1])
    #for state 5, the loop starts from 0
    if (state == 4):
      forward_reward = probAction * (reward_f5a + gamma * next[state])+ probNotAction * (reward_b + gamma * next[0])
      backward_reward = probAction * (reward_b + gamma * next[0])+ probNotAction * (reward_f5a + gamma * next[state])
    return max(forward_reward, backward_reward)

#An inital value iteration is made
reward = initial_val
for i in range(0,5):
  reward[i] = max_value_action_2(i, initial_val)
print("After 1st Iteration:")
print(reward)

#start the value iteration for total_iteration times
def ValueIteration(total_iteration):
  final_reward = initial_val
  reward_p = reward #from 1st iteration
  for _ in range(total_iteration-1):
    for i in range(0, len(states)):
      final_reward[i] = max_value_action_2(i, reward_p)
    reward_p = final_reward
  return final_reward

final_value = ValueIteration(1000)
print("Value after 1000 Interations ::")
print(final_value)

After 1st Iteration:
[1.6, 2.8160000000000003, 2.8160000000000003, 2.8160000000000003, 8.704]
Value after 1000 Interations ::
[61.37948159999994, 64.89128959999994, 69.51208959999994, 75.59208959999994, 83.59208959999992]


- Value obtained from from 3 ( **Q-learning**)::

        [[62.73119814, 61.61184087],
       [66.77077242, 61.8414113 ],
       [71.61108039, 62.89457625],
       [76.54452395, 62.27858727],
       [81.55990703, 67.58229107]

- Value obtained from 4a ( **value Interation** )::

    [61.37948159999994, 

    64.89128959999994, 

    69.51208959999994,

    75.59208959999994,
    
    83.59208959999992]

- the result of both simple q-learning (3a) and  value iteration algoritm (4a) look almost the same and the slight difference in the values we belive could be beacause of default value of parameters used in 3a (for exploration and exploitation). Also, the total number of interations for learning rate could also make a difference


**4b)** What is the importance of exploration in RL? Explain with an example.

**[Answer]** Exploration is an important component of reinforcement learning (RL). As reinforcement learning scales up to address increasingly complex problems, efficient exploration is increasingly important.


Exploration is important in order to find out new actions that could in the long run lead to higher rewards. A real life example could be when choosing which restaurant to go to. One option is a restaurant that you are familiar with which you always enjoy. The second option is another restaurant that you have heard serve very tasty food. Now this leaves you with the choice to either go to the first restaurant that you know will leave you satisfied(**exploit**) or try the other restaurant(**explore**) that might even end up being better than the first one. Do you risk trying the new restaurant? the outcomes could be that you are not as satisfied as when you visit the first restaurant or another outcome could be that you found a better restaurant and now you will be visiting this one instead of the first one, which in long term will be helpful because now you have found a better restaurant that you would not have found if you did not decide to explore.

## Question 5

**5a)** Give a summary of how a decision tree works and how it extends to random forests.

**[Answer]** A decision tree is a class of *regression and classification* tasks which can be represented with a *tree* structured graph. Decision tree algorithm belongs to the family of supervised learning algorithms, used for solving regression and classification problems.

The nodes, branches and leaves in a decision tree represent questions, answers, labels respectively, starting with the **root** (the 1st question) of the tree. The creation of sub-nodes increases the homogeneity of resultant sub-nodes. For each level in the tree, a new question is asked and subsequently, a new branch is added and a new area in the feature space is split off. We continue doing this until a set number of splits has been achieved. A decision tree splits the nodes on all available variables and accordingly selects the split, resulting in most homogeneous sub-nodes, which affects a tree’s accuracy. Decision trees use multiple algorithms/methods to decide to split a node into two or more sub-nodes. The purity of the node increases with respect to the target variable and the algorithm selection is also based on the type of target variables.

**5b)** Explain what makes reinforcement learning different from supervised learning tasks such as regression or classification.

**[Answer]** Supervised learning is when a model learns from a labeled data-set with some guidance and the Reinforcement-Learning is when a machine or an agent interacts with its environment, performs actions, and learns by a trial-and-error method i.e. the RL agent doens't know it's environment, but can interact with it to learn and improve. 

Comparing Supervised learning tasks such as *regression or classification*,  as the model learns from a labeled data-set i.e. we will need to have the previous understanding of the environment so as to get the best performance and also doesn't include exploration. In Reinforcement-Learning, it is approached by *learns by a trial-and-error method*, making is more independent approach as compared to supervised learning approach. The RL learning agent interacts with
the environment by executing actions and learns from errors/rewards.





# References
Primer/text based on the following references:
* http://www.cse.chalmers.se/~chrdimi/downloads/book.pdf
* https://github.com/olethrosdc/ml-society-science/blob/master/notes.pdf